src/Core/Settings.h

```diff
@@ -302,6 +302,7 @@ class IColumn;
     M(Bool, http_skip_not_found_url_for_globs, true, "Skip url's for globs with HTTP_NOT_FOUND error", 0) \
     M(Bool, optimize_throw_if_noop, false, "If setting is enabled and OPTIMIZE query didn't actually assign a merge then an explanatory exception is thrown", 0) \
     M(Bool, use_index_for_in_with_subqueries, true, "Try using an index if there is a subquery or a table expression on the right side of the IN operator.", 0) \
+    M(UInt64, use_index_for_in_with_subqueries_max_values, 0, "Don't use index of a table for filtering by right hand size of the IN operator if the size of set is larger than specified threshold. This allows to avoid performance degradation and higher memory usage due to preparation of additional data structures.", 0) \
```

Comment:
```suggestion
    M(UInt64, max_set_size_to_use_table_index, 0, "The maximal size of set in the right hand side of the IN operator to use table index for filtering. It allows to avoid performance degradation and higher memory usage due to preparation of additional data structures for large queries. Zero means no limit.", 0) \
```

Reply:
This setting was resurrected from some old PR https://github.com/ClickHouse/ClickHouse/issues/39865
It's name complements the boolean setting `use_index_for_in_with_subqueries` a line above. Maybe we should keep the name and only update the description?

Reply:
But these settings have different types, so `use_...` is suitable for a boolean but `max_...` highlight that it's numeric. We can also go with `use_index_for_in_with_subqueries_set_max_size` to something like.
But anyway, any option is ok, no strong opinion on that.